src/sage/rings/species.py

```diff
@@ -0,0 +1,1268 @@
+
+    def _repr_(self):
+        r"""
+        Return a string representation of ``self``.
+
+        TESTS::
+
+            sage: AtomicSpecies("X")
+            Atomic species in X
+            sage: AtomicSpecies("X, Y")
+            Atomic species in X, Y
+        """
+        if len(self._names) == 1:
+            return f"Atomic species in {self._names[0]}"
+        return f"Atomic species in {', '.join(self._names)}"
+
+    Element = AtomicSpeciesElement
+
+
+class MolecularSpecies(IndexedFreeAbelianMonoid, ElementCache):
+    @staticmethod
+    def __classcall__(cls, indices, prefix, **kwds):
+        return super(IndexedMonoid, cls).__classcall__(cls, indices, prefix, **kwds)
+
+    def __init__(self, indices, prefix, **kwds):
+        category = Monoids() & InfiniteEnumeratedSets()
+        IndexedFreeAbelianMonoid.__init__(self, indices, prefix=prefix, category=category, **kwds)
+        ElementCache.__init__(self)
+        self._k = indices._k
+
+    def _project(self, G, pi, part):
+        r"""
+        Project `G` onto a subset ``part`` of its domain.
+
+        ``part`` must be a union of cycles, but this is not checked.
+        """
+        restricted_gens = [[cyc for cyc in gen.cycle_tuples() if cyc[0] in part] for gen in G.gens()]
+        mapping = dict()
+        for k, v in pi.items():
+            es = [e for e in v if e in part]
+            if es:
+                mapping[k] = es
+        return PermutationGroup(gens=restricted_gens, domain=part), mapping
+
+    def _element_constructor_(self, G, pi=None):
+        r"""
+        Construct the `k`-variate molecular species with the given data.
+
+        INPUT:
+
+        - ``G`` - an element of ``self`` (in this case pi must be ``None``)
+          or a permutation group, or a pair ``(X, a)`` consisting of a
+          finite set and an action.
+        - ``pi`` - a dict mapping sorts to iterables whose union is the
+          domain of ``G`` (if ``G`` is a permutation group) or `X` (if ``G``)
+          is a pair ``(X, a)``. If `k=1`, `pi` can be omitted.
+
+        If `G = (X, a)`, then `X` should be a finite set and `a` a transitive
+        action of `G` on `X`.
+
+        EXAMPLES::
+
+            sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: P(SymmetricGroup(4).young_subgroup([2, 2]), {1: [1,2], 2: [3,4]})
+            E_2(X)*E_2(Y)
+
+            sage: X = SetPartitions(4, 2)
+            sage: a = lambda g, x: SetPartition([[g(e) for e in b] for b in x])
+            sage: P((X, a), {1: [1,2], 2: [3,4]})
+            X^2*E_2(Y) + X^2*Y^2 + E_2(X)*Y^2 + E_2(X)*E_2(Y)
+
+        TESTS::
+
+            sage: P = PolynomialSpecies(ZZ, ["X", "Y"])
+            sage: M = P._indices
+            sage: M(CyclicPermutationGroup(4), {1: [1,2], 2: [3,4]})
+            Traceback (most recent call last):
+            ...
+            ValueError: For each orbit of Cyclic group of order 4 as a permutation group, all elements must belong to the same sort
```

Comment:
This is wonderful!  However, the examples should go to the docstring of `PolynomialSpecies._element_constructor_`.

Other than that: is the input checking necessary here?  I imagine it would be caught by the `AtomicSpecies._element_constructor_` anyway.